WIP: fix interaction of software requirements and environments (#1378)
common-workflow-language/cwltool: README.rst

This file may be specified as either XML or YAML and very simply describes various
plugins to enable to "resolve" ``SoftwareRequirement`` dependencies.

Using these hints will allow cwltool to modify the environment in

> which your tool runs, for example by loading one or more environment
> modules. The enviroment is constructed as normal (i.e. standard CWL

Reviewer: ```suggestion
which your tool runs, for example by loading one or more `environment modules
<https://en.wikipedia.org/wiki/Environment_Modules_(software)>`_. The environment
is constructed as normal (i.e. standard CWL
```

runtime enviroment, ``EnvVarRequirement``, ``--preserve-environment``

> are applied), then this is modified. This currently means that you

Reviewer: ```suggestion
are applied), then the environment may modified by the selected tool resolver.  This currently means that you
```

> cannot override variables set in this way.

Reviewer: ```suggestion
cannot override any environment variables set by the selected tool resolver.
```

  cwltool --print-rdf --rdf-serializer=turtle mywf.cwl

Environment Variables in cwltool
--------------------------------

This reference implementation supports several ways of setting
enviroment variables for tools, in addition to the standard
``EnvVarRequirement``. The sequence of steps applied to create the
enviroment is:

> 0. If ``--preserve-entire-environment`` is set, begin with the current

Reviewer: ```suggestion
0. If the ``--preserve-entire-environment`` flag is present, then begin with the current
```

   environment, else begin with an empty environment.

1. Add any variables specified by ``--preserve-environment`` option(s).

> 2. Set ``TMPDIR`` and ``HOME`` per the CWL specification.

Reviewer: ```suggestion
2. Set ``TMPDIR`` and ``HOME`` per `the CWL v1.0+ CommandLineTool specification <https://www.commonwl.org/v1.0/CommandLineTool.html#Runtime_environment>`_.
```

> 3. Apply ``EnvVarRequirement`` from tool description.

Reviewer: ```suggestion
3. Apply any ``EnvVarRequirement`` from the ``CommandLineTool`` description.
```

> 4. Apply any manipulations required by ``MPIRequirement`` extension.

Reviewer: ```suggestion
4. Apply any manipulations required by any ``cwltool:MPIRequirement`` extensions.
```

> 5. Substitute any secrets required by ``Secrets`` extension.
> 
> 6. Modify the environment in response to ``SoftwareRequirement``.

Reviewer: ```suggestion
5. Substitute any secrets required by ``Secrets`` extension.

6. Modify the environment in response to ``SoftwareRequirement`` (see below).
```
